File: python/alethia/assess/metrics.py
"""Label-free embedding quality metrics for entity matching."""


import numpy as np

from ..embedder import l2_normalize, rank_key, top_k_stable


def _ensure_normalized(arr: np.ndarray, *, normalized: bool = False) -> np.ndarray:
    """Return a float64 normalized matrix, trusting an internal normalized input."""
    if normalized:
        result = np.asarray(arr)
        if result.dtype != np.float64:
            raise TypeError("pre-normalized metric inputs must have dtype float64")
        return result
    return l2_normalize(arr)




#: 4096 x n float64 stays under a GB
_BLOCK_ROWS = 4096
#: sorting and partitioning copy their input, so blocks stay near 20 MB
_RETRIEVAL_BLOCK_ROWS = 128

def _nearest_other_similarity(emb: np.ndarray, block_rows: int = _BLOCK_ROWS) -> np.ndarray:
    """Cosine similarity from each row to its nearest *other* row."""
    n = emb.shape[0]
    out = np.empty(n, dtype=np.float64)
    for start in range(0, n, block_rows):
        stop = min(start + block_rows, n)
        block = emb[start:stop] @ emb.T  # (rows, n)
        block[np.arange(stop - start), np.arange(start, stop)] = -np.inf
        out[start:stop] = block.max(axis=1)
    return out


def reference_separability(
    ref_emb: np.ndarray, confusion_threshold: float = 0.9, *, _normalized: bool = False
) -> dict[str, float]:
    """Nearest-other-reference cosine statistics: mean, median, and confusability rate.

    Lower is better on all three. Raw cosines, so they partly reflect anisotropy
    (Ethayarajh 2019); see :func:`centered_separability` for the corrected view.
    """
    ref_emb = _ensure_normalized(ref_emb, normalized=_normalized)
    n = ref_emb.shape[0]
    if n < 2:
        return {
            "mean_nn_similarity": 0.0,
            "median_nn_similarity": 0.0,
            "confusability_rate": 0.0,
        }
    nn_sim = _nearest_other_similarity(ref_emb)
    return {
        "mean_nn_similarity": float(np.mean(nn_sim)),
        "median_nn_similarity": float(np.median(nn_sim)),
        "confusability_rate": float(np.mean(nn_sim > confusion_threshold)),
    }


def centered_separability(ref_emb: np.ndarray) -> dict[str, float]:
    """Anisotropy-corrected separability: nearest-neighbour margin after mean removal.

    Returns ``centered_nn_similarity`` (lower better) and ``nn_margin_z``, how many
    standard deviations the top neighbour stands above a typical one (scale-free).

    References:
        Ethayarajh (2019), doi:10.48550/arXiv.1909.00512. Su et al. (2021),
        doi:10.48550/arXiv.2103.15316.
    """
    emb = np.asarray(ref_emb, dtype=np.float64)
    n = emb.shape[0]
    if n < 3:
        return {"centered_nn_similarity": 0.0, "nn_margin_z": 0.0}

    centered = l2_normalize(emb - emb.mean(axis=0, keepdims=True))

    # one blocked pass produces all three accumulators
    nn_sim = np.empty(n, dtype=np.float64)
    sums = np.empty(n, dtype=np.float64)
    sqsums = np.empty(n, dtype=np.float64)
    for start in range(0, n, _BLOCK_ROWS):
        stop = min(start + _BLOCK_ROWS, n)
        block = centered[start:stop] @ centered.T
        rows, cols = np.arange(stop - start), np.arange(start, stop)
        # self-similarity is exactly 1.0 on normalized rows
        sums[start:stop] = block.sum(axis=1) - 1.0
        # avoids materializing block**2
        sqsums[start:stop] = np.einsum("ij,ij->i", block, block) - 1.0
        block[rows, cols] = -np.inf
        nn_sim[start:stop] = block.max(axis=1)

    m = n - 1
    mu = sums / m
    var = np.maximum(0.0, sqsums / m - mu**2)
    sd = np.sqrt(var)
    sd[sd == 0] = np.nan  # degenerate rows drop out of the nanmean below
    margin_z = (nn_sim - mu) / sd

    return {
        "centered_nn_similarity": float(np.mean(nn_sim)),
        "nn_margin_z": float(np.nanmean(margin_z)),
    }


def intrinsic_dimensionality(emb: np.ndarray) -> dict[str, float]:
    """Effective dimensionality as a PCA participation ratio.

    Returns ``participation_ratio`` and ``normalized_pr``, its ratio to the ambient dim.
    """
    emb = np.asarray(emb, dtype=np.float64)
    if emb.shape[0] < 2:
        return {"participation_ratio": 0.0, "normalized_pr": 0.0}
    centered = emb - emb.mean(axis=0, keepdims=True)
    # the d x d form has the same nonzero eigenvalues as the n x n one
    lam = np.linalg.eigvalsh(centered.T @ centered)
    lam = np.maximum(lam, 0.0)
    total = np.sum(lam)
    if total <= 0:
        return {"participation_ratio": 0.0, "normalized_pr": 0.0}
    pr = (total**2) / np.sum(lam**2)
    return {
        "participation_ratio": float(pr),
        "normalized_pr": float(pr / emb.shape[1]),
    }




def alignment_loss(
    src_emb: np.ndarray, var_emb: np.ndarray, alpha: float = 2.0
) -> float:
    """Alignment: mean distance^alpha between positive pairs (lower = better)."""
    if src_emb.shape[0] == 0:
        return float("nan")
    a = l2_normalize(src_emb)
    b = l2_normalize(var_emb)
    d = np.linalg.norm(a - b, axis=1)
    return float(np.mean(d**alpha))


def positive_pair_rank(
    src_emb: np.ndarray, var_emb: np.ndarray, block_rows: int = _BLOCK_ROWS
) -> float:
    """Mean percentile rank of each positive variant among all variants, ties mid-ranked."""
    if src_emb.shape[0] == 0 or var_emb.shape[0] == 0:
        return float("nan")
    if src_emb.shape[0] != var_emb.shape[0]:
        raise ValueError("src_emb and var_emb must have the same number of rows")

    src = l2_normalize(src_emb)
    var = l2_normalize(var_emb)
    n = src.shape[0]
    ranks = np.empty(n, dtype=np.float64)
    for start in range(0, n, block_rows):
        stop = min(start + block_rows, n)
        sims = src[start:stop] @ var.T
        positive = sims[np.arange(stop - start), np.arange(start, stop)]
        # rounded, so ties compare equal on every BLAS build
        sims = rank_key(sims)
        positive = rank_key(positive)
        lower = np.sum(sims < positive[:, None], axis=1)
        equal = np.sum(sims == positive[:, None], axis=1)
        ranks[start:stop] = (lower + 0.5 * equal) / n
    return float(np.mean(ranks))


def uniformity_loss(
    emb: np.ndarray, t: float = 2.0, max_points: int = 2000, *, _normalized: bool = False
) -> float:
    """Uniformity ``log E exp(-t * ||x-y||^2)`` over pairs; more negative is better.

    After Wang & Isola (2020), doi:10.48550/arXiv.2005.10242.
    """
    n = emb.shape[0]
    if n < 2:
        return float("nan")
    if n > max_points:
        idx = np.linspace(0, n - 1, max_points).astype(int)
        emb = emb[idx]
    emb = _ensure_normalized(emb, normalized=_normalized)
    # ||x - y||^2 = 2 - 2 x.y on unit rows, so no distance array is built
    gram = emb @ emb.T
    total = 0.0
    pairs = 0
    for row in range(emb.shape[0] - 1):
        pdist = np.maximum(0.0, 2.0 - 2.0 * gram[row, row + 1 :])
        total += float(np.exp(-t * pdist).sum())
        pairs += pdist.size
    return float(np.log(total / pairs))




def _query_ref_sims(
    query_emb: np.ndarray, ref_emb: np.ndarray, *, normalized: bool = False
) -> np.ndarray:
    return (_ensure_normalized(query_emb, normalized=normalized)
            @ _ensure_normalized(ref_emb, normalized=normalized).T)


def retrieval_margin(
    query_emb: np.ndarray, ref_emb: np.ndarray, low_margin_z: float = 1.0, *,
    _normalized: bool = False, _sims: np.ndarray | None = None,
) -> dict[str, float]:
    """Top-1 vs top-2 margin per query, in raw cosine and in standard deviations.

    ``mean_margin`` / ``low_margin_rate`` are absolute cosines and so confounded by
    anisotropy; the ``_z`` variants divide by the per-query spread and compare fairly
    across models, which is why the composite weights those.
    """
    sims = (_query_ref_sims(query_emb, ref_emb, normalized=_normalized)
            if _sims is None else _sims)
    if sims.shape[1] < 2:
        return {
            "mean_margin": 0.0,
            "low_margin_rate": 1.0,
            "mean_margin_z": 0.0,
            "low_margin_z_rate": 1.0,
        }
    margin = np.empty(sims.shape[0], dtype=np.float64)
    sd = np.empty(sims.shape[0], dtype=np.float64)
    for start in range(0, sims.shape[0], _RETRIEVAL_BLOCK_ROWS):
        stop = min(start + _RETRIEVAL_BLOCK_ROWS, sims.shape[0])
        block = sims[start:stop]
        part = np.partition(block, -2, axis=1)
        margin[start:stop] = part[:, -1] - part[:, -2]
        sd[start:stop] = block.std(axis=1)
    sd = np.where(sd == 0, np.nan, sd)
    margin_z = margin / sd

    return {
        "mean_margin": float(np.mean(margin)),
        "low_margin_rate": float(np.mean(margin < 0.02)),
        "mean_margin_z": float(np.nanmean(margin_z)),
        "low_margin_z_rate": float(
            np.nanmean(np.where(np.isnan(margin_z), np.nan, margin_z < low_margin_z))
        ),
    }


def hubness(
    query_emb: np.ndarray, ref_emb: np.ndarray, k: int = 5, *,
    _normalized: bool = False, _sims: np.ndarray | None = None,
) -> dict[str, float]:
    """Skewness of the k-occurrence distribution; high means pathological hubs.

    After Radovanovic et al. (2010), JMLR 11:2487-2531.
    """
    q = _ensure_normalized(query_emb, normalized=_normalized)
    r = _ensure_normalized(ref_emb, normalized=_normalized)
    n_ref = r.shape[0]
    k = min(k, n_ref)
    if k < 1:
        return {"hubness_skew": 0.0}
    counts = np.zeros(n_ref, dtype=np.float64)
    for start in range(0, q.shape[0], _RETRIEVAL_BLOCK_ROWS):
        stop = min(start + _RETRIEVAL_BLOCK_ROWS, q.shape[0])
        sims = q[start:stop] @ r.T if _sims is None else _sims[start:stop]
        topk_idx = top_k_stable(sims, k)
        counts += np.bincount(topk_idx.ravel(), minlength=n_ref)
    mu, sigma = counts.mean(), counts.std()
    if sigma == 0:
        return {"hubness_skew": 0.0}
    skew = float(np.mean(((counts - mu) / sigma) ** 3))
    return {"hubness_skew": skew}


def mutual_nn_rate(
    query_emb: np.ndarray, ref_emb: np.ndarray, k: int = 5, *,
    _normalized: bool = False, _sims: np.ndarray | None = None,
) -> float:
    """Fraction of query->ref top-1 assignments that are reciprocated within ref top-k."""
    q = _ensure_normalized(query_emb, normalized=_normalized)
    r = _ensure_normalized(ref_emb, normalized=_normalized)
    nq, nr = q.shape[0], r.shape[0]
    if nq == 0 or nr == 0:
        return float("nan")
    k = min(k, nq)

    # blocked so the (nq, nr) matrix is never held whole
    q_top1 = np.empty(nq, dtype=np.intp)
    for start in range(0, nq, _BLOCK_ROWS):
        stop = min(start + _BLOCK_ROWS, nq)
        block = q[start:stop] @ r.T if _sims is None else _sims[start:stop]
        q_top1[start:stop] = np.argmax(rank_key(block), axis=1)

    # only distinct chosen references need a query row, avoiding an (nq, nq) gather
    chosen_refs, inverse = np.unique(q_top1, return_inverse=True)
    topk = np.empty((len(chosen_refs), k), dtype=np.intp)
    # the _sims branch gathers columns, copying (nq, block)
    for start in range(0, len(chosen_refs), _RETRIEVAL_BLOCK_ROWS):
        stop = min(start + _RETRIEVAL_BLOCK_ROWS, len(chosen_refs))
        block = (r[chosen_refs[start:stop]] @ q.T if _sims is None
                else _sims[:, chosen_refs[start:stop]].T)
        topk[start:stop] = top_k_stable(block, k)

    # topk[inverse] is (nq, k): each query's chosen reference's top-k queries
    return float((topk[inverse] == np.arange(nq)[:, None]).any(axis=1).mean())

File: web/src/metrics.js
// Label-free geometric metrics, ported from alethia.assess.metrics. Each states its
// formula so the three implementations can be checked against one another.

import { l2Normalize, mean, median, popStd, rankKey, topK } from './core.js';

function simsRow(x, i, y) {
  const out = new Float64Array(y.rows);
  const xo = i * x.cols;
  for (let j = 0; j < y.rows; j++) {
    const yo = j * y.cols;
    let sum = 0;
    for (let k = 0; k < x.cols; k++) sum += x.data[xo + k] * y.data[yo + k];
    out[j] = sum;
  }
  return out;
}

/**
 * How well distinct references are kept apart: cosine to the nearest *other* reference.
 *
 * Raw cosines, so they partly reflect anisotropy: transformer embeddings tend to occupy
 * a narrow cone that pushes every cosine high. See centeredSeparability for the
 * corrected view.
 */
export function referenceSeparability(refEmb, confusionThreshold = 0.9) {
  const x = l2Normalize(refEmb);
  const n = x.rows;
  if (n < 2) {
    return { mean_nn_similarity: 0, median_nn_similarity: 0, confusability_rate: 0 };
  }
  const nn = new Float64Array(n);
  for (let i = 0; i < n; i++) {
    const s = simsRow(x, i, x);
    s[i] = -Infinity;
    let best = -Infinity;
    for (let j = 0; j < n; j++) if (s[j] > best) best = s[j];
    nn[i] = best;
  }
  let confusable = 0;
  for (const v of nn) if (v > confusionThreshold) confusable++;
  return {
    mean_nn_similarity: mean(nn),
    median_nn_similarity: median(nn),
    confusability_rate: confusable / n,
  };
}

/** Separability after removing the set mean, which strips the shared cone direction. */
export function centeredSeparability(refEmb) {
  const n = refEmb.rows;
  const d = refEmb.cols;
  if (n < 3) return { centered_nn_similarity: 0, nn_margin_z: 0 };

  const centered = new Float64Array(refEmb.data.length);
  for (let j = 0; j < d; j++) {
    let mu = 0;
    for (let i = 0; i < n; i++) mu += refEmb.data[i * d + j];
    mu /= n;
    for (let i = 0; i < n; i++) centered[i * d + j] = refEmb.data[i * d + j] - mu;
  }
  const x = l2Normalize({ data: centered, rows: n, cols: d });

  const nn = new Float64Array(n);
  const z = new Float64Array(n);
  for (let i = 0; i < n; i++) {
    const s = simsRow(x, i, x);
    // Over the n-1 other references; a reference is not its own neighbour.
    const others = new Float64Array(n - 1);
    let w = 0;
    let best = -Infinity;
    for (let j = 0; j < n; j++) {
      if (j === i) continue;
      others[w++] = s[j];
      if (s[j] > best) best = s[j];
    }
    nn[i] = best;
    const mu = mean(others);
    const sd = popStd(others);
    // A degenerate row has no scale to express a margin in, so it drops out.
    z[i] = sd > 0 ? (best - mu) / sd : NaN;
  }
  const defined = Array.from(z).filter((v) => !Number.isNaN(v));
  return {
    centered_nn_similarity: mean(nn),
    nn_margin_z: defined.length ? mean(defined) : NaN,
  };
}

/**
 * Eigenvalues of a symmetric matrix by the cyclic Jacobi method.
 *
 * Runs only on the d-by-d feature covariance, never the n-by-n similarity matrix, so it
 * stays small however much data the user pastes in.
 */
export function symmetricEigenvalues(a, d, { sweeps = 60, tol = 1e-12 } = {}) {
  const m = Float64Array.from(a);
  const at = (i, j) => m[i * d + j];
  const set = (i, j, v) => {
    m[i * d + j] = v;
  };
  for (let sweep = 0; sweep < sweeps; sweep++) {
    let off = 0;
    for (let p = 0; p < d; p++) {
      for (let q = p + 1; q < d; q++) off += at(p, q) ** 2;
    }
    if (off <= tol) break;
    for (let p = 0; p < d; p++) {
      for (let q = p + 1; q < d; q++) {
        const apq = at(p, q);
        if (Math.abs(apq) < 1e-300) continue;
        const theta = (at(q, q) - at(p, p)) / (2 * apq);
        const t = Math.sign(theta || 1) / (Math.abs(theta) + Math.sqrt(theta * theta + 1));
        const c = 1 / Math.sqrt(t * t + 1);
        const s = t * c;
        for (let k = 0; k < d; k++) {
          const akp = at(k, p);
          const akq = at(k, q);
          set(k, p, c * akp - s * akq);
          set(k, q, s * akp + c * akq);
        }
        for (let k = 0; k < d; k++) {
          const apk = at(p, k);
          const aqk = at(q, k);
          set(p, k, c * apk - s * aqk);
          set(q, k, s * apk + c * aqk);
        }
      }
    }
  }
  const eigenvalues = new Float64Array(d);
  for (let i = 0; i < d; i++) eigenvalues[i] = at(i, i);
  return eigenvalues;
}

/**
 * Participation ratio over the PCA eigenvalues, and the same divided by the ambient
 * dimension. Higher means the model spreads data over more directions than it collapses.
 */
export function intrinsicDimensionality(emb) {
  const n = emb.rows;
  const d = emb.cols;
  if (n < 2) return { participation_ratio: 0, normalized_pr: 0 };

  const centered = new Float64Array(emb.data.length);
  for (let j = 0; j < d; j++) {
    let mu = 0;
    for (let i = 0; i < n; i++) mu += emb.data[i * d + j];
    mu /= n;
    for (let i = 0; i < n; i++) centered[i * d + j] = emb.data[i * d + j] - mu;
  }
  // shares the nonzero eigenvalues of the n-by-n Gram matrix and is far smaller
  const cov = new Float64Array(d * d);
  for (let p = 0; p < d; p++) {
    for (let q = p; q < d; q++) {
      let sum = 0;
      for (let i = 0; i < n; i++) sum += centered[i * d + p] * centered[i * d + q];
      cov[p * d + q] = sum;
      cov[q * d + p] = sum;
    }
  }
  const lam = symmetricEigenvalues(cov, d);
  let total = 0;
  let sqTotal = 0;
  for (let i = 0; i < d; i++) {
    const v = Math.max(lam[i], 0);
    total += v;
    sqTotal += v * v;
  }
  if (total <= 0) return { participation_ratio: 0, normalized_pr: 0 };
  const pr = (total * total) / sqTotal;
  return { participation_ratio: pr, normalized_pr: pr / d };
}

/** Mean distance between matched positive pairs; lower means dirt moves a string less. */
export function alignmentLoss(srcEmb, varEmb, alpha = 2) {
  const a = l2Normalize(srcEmb);
  const b = l2Normalize(varEmb);
  if (!a.rows || a.rows !== b.rows) return NaN;
  const out = new Float64Array(a.rows);
  for (let i = 0; i < a.rows; i++) {
    let sum = 0;
    for (let j = 0; j < a.cols; j++) sum += (a.data[i * a.cols + j] - b.data[i * b.cols + j]) ** 2;
    out[i] = Math.sqrt(Math.max(0, sum)) ** alpha;
  }
  return mean(out);
}

/** log E exp(-t d^2) over all pairs; more negative means a better-spread space. */
export function uniformityLoss(emb, t = 2, maxPoints = 2000) {
  let x = emb;
  let n = x.rows;
  if (n < 2) return NaN;
  if (n > maxPoints) {
    // Even stride, matching the linspace subsample elsewhere.
    const idx = new Int32Array(maxPoints);
    for (let i = 0; i < maxPoints; i++) {
      idx[i] = Math.trunc((i * (n - 1)) / (maxPoints - 1));
    }
    const data = new Float64Array(maxPoints * x.cols);
    for (let i = 0; i < maxPoints; i++) {
      data.set(x.data.subarray(idx[i] * x.cols, (idx[i] + 1) * x.cols), i * x.cols);
    }
    x = { data, rows: maxPoints, cols: x.cols };
    n = maxPoints;
  }
  x = l2Normalize(x);
  let total = 0;
  let pairs = 0;
  for (let i = 0; i < n - 1; i++) {
    const s = simsRow(x, i, x);
    for (let j = i + 1; j < n; j++) {
      // For unit rows ||x-y||^2 = 2 - 2 x.y, so no difference vector is ever formed.
      total += Math.exp(-t * Math.max(0, 2 - 2 * s[j]));
      pairs++;
    }
  }
  return Math.log(total / pairs);
}

/** How decisively the top reference beats the runner-up, raw and in per-query sigmas. */
export function retrievalMargin(queryEmb, refEmb, lowMargin = 0.02) {
  const q = l2Normalize(queryEmb);
  const r = l2Normalize(refEmb);
  if (r.rows < 2) {
    // no runner-up, so no margin; Python reports these constants, not NaN
    return { mean_margin: 0, low_margin_rate: 1, mean_margin_z: 0, low_margin_z_rate: 1 };
  }
  if (!q.rows) {
    return {
      mean_margin: NaN, low_margin_rate: NaN, mean_margin_z: NaN, low_margin_z_rate: NaN,
    };
  }
  const margins = new Float64Array(q.rows);
  const zs = new Float64Array(q.rows);
  for (let i = 0; i < q.rows; i++) {
    const s = simsRow(q, i, r);
    const best = topK(s, 2);
    const margin = s[best[0]] - s[best[1]];
    margins[i] = margin;
    const sd = popStd(s);
    zs[i] = sd > 0 ? margin / sd : NaN;
  }
  let low = 0;
  let lowZ = 0;
  for (let i = 0; i < margins.length; i++) {
    if (margins[i] < lowMargin) low++;
    // NaN < 1 is false, so an undefined z is unknown, not low; it stays in the denominator
    if (zs[i] < 1) lowZ++;
  }
  const definedZ = Array.from(zs).filter((v) => !Number.isNaN(v));
  return {
    mean_margin: mean(margins),
    low_margin_rate: low / margins.length,
    mean_margin_z: definedZ.length ? mean(definedZ) : NaN,
    low_margin_z_rate: lowZ / zs.length,
  };
}

/**
 * Skewness of the k-occurrence distribution: are a few references absorbing most
 * queries? High values are pathological. After Radovanovic et al. (2010).
 */
export function hubness(queryEmb, refEmb, k = 5) {
  const q = l2Normalize(queryEmb);
  const r = l2Normalize(refEmb);
  const nRef = r.rows;
  const kk = Math.min(k, nRef);
  if (kk < 1 || !q.rows) return { hubness_skew: 0 };
  const counts = new Float64Array(nRef);
  for (let i = 0; i < q.rows; i++) {
    for (const j of topK(simsRow(q, i, r), kk)) counts[j]++;
  }
  const mu = mean(counts);
  const sigma = popStd(counts);
  if (sigma === 0) return { hubness_skew: 0 };
  let acc = 0;
  for (const c of counts) acc += ((c - mu) / sigma) ** 3;
  return { hubness_skew: acc / counts.length };
}

/** Fraction of top-1 query-to-reference picks that the reference reciprocates. */
export function mutualNnRate(queryEmb, refEmb, k = 5) {
  const q = l2Normalize(queryEmb);
  const r = l2Normalize(refEmb);
  if (!q.rows || !r.rows) return NaN;
  const kk = Math.min(k, q.rows);
  const top1 = new Int32Array(q.rows);
  for (let i = 0; i < q.rows; i++) top1[i] = topK(simsRow(q, i, r), 1)[0];

  // Only the references actually chosen need their query row computed.
  const chosen = [...new Set(Array.from(top1))];
  const backTopK = new Map();
  for (const ref of chosen) {
    backTopK.set(ref, new Set(topK(simsRow(r, ref, q), kk)));
  }
  let reciprocated = 0;
  for (let i = 0; i < q.rows; i++) {
    if (backTopK.get(top1[i]).has(i)) reciprocated++;
  }
  return reciprocated / q.rows;
}

/**
 * Percentile rank of each synthesized positive among all variants, averaged.
 *
 * `src[i]` and `variant[i]` are a positive pair. Unlike alignmentLoss this uses only
 * the *ordering* of each source's similarities, so it is invariant to any strictly
 * increasing transformation of cosine, including the affine shift from adding a
 * common cone direction and renormalising. That invariance is why it carries the
 * largest single weight in the composite. Ties take their mid-rank.
 */
export function positivePairRank(srcEmb, varEmb) {
  if (!srcEmb.rows || !varEmb.rows) return NaN;
  if (srcEmb.rows !== varEmb.rows) {
    throw new Error('srcEmb and varEmb must have the same number of rows');
  }
  const src = l2Normalize(srcEmb);
  const variant = l2Normalize(varEmb);
  const n = src.rows;
  const ranks = new Float64Array(n);
  for (let i = 0; i < n; i++) {
    const sims = simsRow(src, i, variant);
    // an exact == misses equal similarities, shifting mid-ranks between runtimes
    const positive = rankKey(sims[i]);
    let lower = 0;
    let equal = 0;
    for (let j = 0; j < n; j++) {
      const v = rankKey(sims[j]);
      if (v < positive) lower++;
      else if (v === positive) equal++;
    }
    ranks[i] = (lower + 0.5 * equal) / n;
  }
  return mean(ranks);
}
